File: smart-hub/src/main/scala/io/github/tassiLuca/hub/infrastructure/SwingDashboard.scala
package io.github.tassiLuca.hub.infrastructure

import io.github.tassiLuca.hub.infrastructure.ui.DashboardUI
import io.github.tassiLuca.hub.core.Temperature
import io.github.tassiLuca.hub.core.ports.DashboardServiceComponent

import javax.swing.SwingUtilities

trait SwingDashboard extends DashboardServiceComponent:
  override val dashboard: DashboardService = new DashboardService:

    private val view = DashboardUI()

    override def temperatureUpdated(temperature: Temperature): Unit = SwingUtilities.invokeLater { () =>
      view.temperatureLabel.setText(s"$temperature °C")
    }

    override def offHeaterNotified(): Unit = SwingUtilities.invokeLater { () =>
      view.heaterLabel.setText("Off")
    }

    override def onHeaterNotified(): Unit = SwingUtilities.invokeLater { () =>
      view.heaterLabel.setText("On")
    }

    override def alertNotified(msg: String): Unit = SwingUtilities.invokeLater { () =>
      view.alertTextArea.setText(msg)
    }

File: smart-hub/src/main/kotlin/io/github/tassiLuca/hubkt/adapters/SwingDashboard.kt
package io.github.tassiLuca.hubkt.adapters

import io.github.tassiLuca.hubkt.adapters.ui.DashboardUI
import io.github.tassiLuca.hubkt.core.Temperature
import io.github.tassiLuca.hubkt.core.ports.DashboardService

/** Swing dashboard. */
class SwingDashboard : DashboardService {
    private val view = DashboardUI()

    override fun temperatureUpdated(newTemperature: Temperature) {
        view.temperatureLabel.setText("$newTemperature °C")
    }

    override fun onHeaterNotified() {
        view.heaterLabel.setText("Off")
    }

    override fun offHeaterNotified() {
        view.heaterLabel.setText("On")
    }

    override fun alertNotified(msg: String) {
        view.alertsModel.insertRow(0, arrayOf(msg))
    }
}
